outlines/inputs.py

```diff
@@ -0,0 +1,179 @@
+    content: str
+
+    def __post_init__(self):
+        if self.role not in ["user", "assistant", "system"]:
+            raise ValueError(
+                "Invalid role. The only valid roles are 'user', 'assistant' "
+                "and 'system'."
+            )
+
+        if not isinstance(self.content, str):
+            raise ValueError("Content must be a string")
+
+
+@dataclass
+class Chat:
+    """Contains the input for a chat model.
+
+    Provide an instance of this class as the `model_input` argument to a model
+    that supports chat.
+
+    On top of `Vision` instances, the messages can be dicts with 'role' and
+    'content' keys. The role can be 'user', 'assistant', or 'system'. The content
+    must be a string.
+
+    Examples
+    --------
+    ```python
+    chat_prompt = Chat([
+        {"role": "system", "content": "You are a helpful assistant."},
+        {"role": "user", "content": "Describe the image below please."},
+        Vision("Title: A beautiful sunset over a calm ocean.", image),
+    ])
+    ```
+
+    Parameters
+    ----------
+    messages
+        A list of messages.
+
+    """
+    messages: List[Union[dict, Message, Vision]]
+
+    def __post_init__(self):
+        processed_messages = []
+        for message in self.messages:
+            processed_messages.append(self._input_processing(message))
+        self.messages = processed_messages
+
+    def _input_processing(
+        self, input: Union[dict, Message, Vision]
+    ) -> Union[Message, Vision]:
+        """Process the user-provided input to return an object that can be
+        added to the message list.
+
+        """
+        if isinstance(input, dict):
+            return Message(**input)
+        elif isinstance(input, Message) or isinstance(input, Vision):
+            return input
+        else:
+            raise ValueError(
+                "Invalid message type. The only valid message types are "
+                "dict, Message and Vision."
+            )
+
+    def add(self, message: Union[dict, Message, Vision], index: int = -1):
+        """Add a message to the chat.
+
+        Parameters
+        ----------
+        message
+            Either a dict with 'role' and 'content' keys, or a `Message` or
+            `Vision` instance.
+        index
+            The index at which to insert the message. If not provided, the
+            message will be appended to the end of the list.
+
+        """
+        self.messages.insert(index, self._input_processing(message))
+
+    def remove(self, index: int = -1):
+        """Remove a message from the chat.
+
+        Parameters
+        ----------
+        index
+            The index of the message to remove. If not provided, the last
+            message will be removed.
+
+        """
+        self.messages.pop(index)
+
+    def __str__(self):
+        return "\n".join(str(message) for message in self.messages)
+
+
+def prompt_string_to_chat(prompt: str) -> Optional[Chat]:
```

Comment:
Using Jinja's parser via [extensions](https://jinja.palletsprojects.com/en/stable/templates/#extensions) would be more robust.

Reply:
Good idea